Move ugly comparison to util

behrmann opened the pull request:
/cc @aafeijoo-suse

behrmann commented on mkosi/util.py:
```diff
@@ -269,3 +269,7 @@ def mandatory_variable(name: str) -> str:
+
+
+def dir_only_contains(dir: Path, tgt: PathString) -> bool:
+    return not dir.is_dir() or (dir / tgt).exists() and len(list(dir.glob("*"))) == 1
```
Thinking about this some more, would it make sense for this the directory to also be skipped if it's empty?

aafeijoo-suse replied:
> Thinking about this some more, would it make sense for this the directory to also be skipped if it's empty?

It would, but... is realistic to think of a system with a `/usr/lib/modules/<kver>` directory with nothing within it? As a curiosity, some yeas ago I had to add a check with `/usr/lib/modules/<kver>/kernel` in dracut to fix a "real" customer case...

https://github.com/openSUSE/dracut/commit/24aa3f7488b0d668902c0d2bc2566e4fdc289ff4

behrmann replied:
> It would, but... is realistic to think of a system with a /usr/lib/modules/<kver> directory with nothing within it? 

That's why I wondered. I mean the Debian system I just checked, (that's long-running and has been updating for over a year now, so quite different from the images we build, has about 20 directories, more than three quarters o which are empty.

septatrix commented on mkosi/util.py:
```diff
@@ -269,3 +269,7 @@ def mandatory_variable(name: str) -> str:
+
+
+def dir_only_contains(dir: Path, tgt: PathString) -> bool:
+    return not dir.is_dir() or (dir / tgt).exists() and len(list(dir.glob("*"))) == 1
```
Instead of the list allocation and eager glob, the following should also do the trick and be more efficient:

```python
try:
    [child] = dir.iterdir()
except ValueError:
    return False # unwrapping failed, either 0 or 2+ children

Now that it is extracted into a function there is no need to keep it a oneliner

behrmann replied:
This is indeed prettier. The whole function becomes a bit unwieldy, but at least it's twice as fast!

aafeijoo-suse commented on mkosi/util.py:
```diff
@@ -269,3 +269,16 @@ def mandatory_variable(name: str) -> str:
         die(f"${name} must be set in the environment")
+
+
+def dir_only_contains(dir: Path, tgt: PathString) -> bool:
```
Shouldn't this be called something like `is_valid_kdir` or so?

`not dir.is_dir():` returing `True` does not mean that `dir` only contains `tgt`...

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -441,3 +441,16 @@ def process_kernel_modules(
+
+
+def is_valid_kdir(kdir: Path) -> bool:
+    if not kdir.is_dir():
+        return True
```
Shouldn't this return False? The function is called is_valid_kdir()

keszybz reviewed (commented):
Oh, I see I didn't press "submit". Sorry for any outdated comments.

keszybz commented on mkosi/util.py:
```diff
@@ -272,4 +272,13 @@ def mandatory_variable(name: str) -> str:
+    try:
+        [child] = dir.iterdir()
+    except ValueError:
+        return False
```
Please ignore the comments about the name. But I think the proposed code is better.

behrmann commented:
Reworked as suggested, though it's a bit unfortunate that the try-except-less code is slower by more than a factor of 2. I added skipping empty directories as well.

septatrix commented:
> Reworked as suggested, though it's a bit unfortunate that the try-except-less code is slower by more than a factor of 2. I added skipping empty directories as well.

It could also be unrolled into two `next(...)` calls to keep it lazy but avoid the try-except (even though the unpacking pattern is a quite common and idiomatic Python...)
